### PR TITLE
Add options to upload / update without publish

### LINKED ISSUE #21: Add support for create without publish
I would like to use the tool in a workflow where I want to use the prereserved DOI before publishing the record. Therefore I need to run create and upload, but not publish.

Working on a PR now, let me know if this would be a welcome feature.

### PR DESCRIPTION
**Pull request by nuest**
Closes #21

### CODE REVIEW

**Comment by nuest** [comment]
AFAICT the failing tests are due to my changes for make pydantic 2 happy, so I'll revert those.

**Comment by nuest** [comment]
Tests work no on my fork, no idea why they do not show up here: https://github.com/nuest/zenodo-client/actions/runs/5892513675

File: src/zenodo_client/api.py [hunk 1]
```diff
@@ -154,48 +165,89 @@ def publish(self, deposition_id: str, sleep: bool = True) -> requests.Response:
...
-    def update(self, deposition_id: str, paths: Paths) -> requests.Response:
-        """Create a new version of the given record with the given files."""
-        # Prepare a new version based on the old version
-        # see: https://developers.zenodo.org/#new-version)
-        res = requests.post(
-            f"{self.depositions_base}/{deposition_id}/actions/newversion",
-            params={"access_token": self.access_token},
-        )
-        res.raise_for_status()
+    def update(
```
Comment (cthoyt) [inline src/zenodo_client/api.py:168]: this function's job is to update the files in a record. I would rather you take all of the new changes that are related to updating the data in a record, and make a new function.

Comment (nuest) [reply src/zenodo_client/api.py:168]: Sure, can do: will remove the parameter `data` and think about a new function. How does `update_metadata()` sound?

Would you also rather keep the function parameters as they are, or are you fine with a new (backwards compatible) option `publish` ?

If files are updated in an existing record, then there will always be a new version, so I can remove `new_version`.

Comment (cthoyt) [reply src/zenodo_client/api.py:168]: update_metadata is great. I'm fine to add the boolean option to turn off publishing for both of these functions

**Comment by cthoyt** [comment]
@nuest any movement forward on this?

**Comment by nuest** [comment]
Hi! Working on this now...

**Comment by nuest** [comment]
Ups, did you just merge by accident?

**Comment by nuest** [comment]
Never mind, must have been a glitch in my local git client.

File: src/zenodo_client/api.py [hunk 2]
```diff
@@ -284,17 +391,38 @@ def download_latest(
         name: str,
         *,
         force: bool = False,
-        parts: PartsHint = None,
+        parts: Optional[PartsHint] = None,
```
Comment (cthoyt) [inline src/zenodo_client/api.py:394]: PartsHint already unions None, you can revert changes like this

**Comment by cthoyt** [comment]
@nuest I did a bit of clean up - linting, adjusting code to reduce diff, removing some minor changes that make it hard to review.

I still need to figure out how to make the environment available for tests so github actions can use my api token

**Comment by cthoyt** [comment]
@nuest I improved the way tox and GHA passes around configuration. Could you please in your fork of the repo add a secret `ZENODO_SANDBOX_API_TOKEN` and put a key from zenodo's sandbox? If you aren't able to get one, I can email you mine.

This will allow the tests to run in full on CI

**Comment by nuest** [comment]
Hi cthoyt!

Sorry for my silence and thanks for your efforts - I have actually continued to work on things myself and now we have to diverged code bases... I will take a look at the token for the CI tomorrow, not sure how to continue with my other branch, see

**https://github.com/nuest/zenodo-client/tree/21-v2**

File: src/zenodo_client/api.py [hunk 3]
```diff
@@ -37,24 +37,34 @@ def ensure_zenodo(key: str, data: Data, paths: Paths, **kwargs) -> requests.Resp
...
+    return Zenodo(**kwargs).publish(deposition_id, sleep=sleep)
 
 
 def download_zenodo(deposition_id: str, name: str, force: bool = False, **kwargs) -> Path:
     """Download a Zenodo record."""
-    return Zenodo(**kwargs).download(deposition_id, name=name, force=force)
+    return Zenodo(**kwargs).download(record_id=record_id, name=name, force=force)
 
 
-def download_zenodo_latest(deposition_id: str, path: str, force: bool = False, **kwargs) -> Path:
```
Comment (cthoyt) [inline src/zenodo_client/api.py:55]: please don't make cosmetic changes like this within a PR that includes other additions and changes

@nuest I also made several reverts before, do you understand why I am asking for this? If not, I can explain further

File: src/zenodo_client/api.py [hunk 4]
```diff
@@ -37,24 +37,34 @@ def ensure_zenodo(key: str, data: Data, paths: Paths, **kwargs) -> requests.Resp
     return Zenodo(**kwargs).ensure(key=key, data=data, paths=paths)
 
 
-def create_zenodo(data: Data, paths: Paths, **kwargs) -> requests.Response:
+def create_zenodo(data: Data, paths: Paths, *, publish: bool = False, **kwargs) -> requests.Response:
```
Comment (cthoyt) [inline src/zenodo_client/api.py:40]: make default True to maintain existing functionality

**Comment by nuest** [comment]
I merged, think I captured all you changes appropriately, but I also made some fixes due to more extensive tests (e.g., fixing upload of files for not-yet-published depositions).

I'll ping you when I'm done with the merge, needs some manual fixes.

File: src/zenodo_client/api.py (hunk 3 above)
Comment (nuest) [reply src/zenodo_client/api.py:55]: No, I understand - I respect you for keeping your codebase and commit history clean like that, something to aspire to.

**Comment by nuest** [comment]
Tests don't run through on my fork, "Too many requests". Maybe there is something wrong with the token?

https://github.com/nuest/zenodo-client/actions/runs/6177578545/job/16768995002

**Comment by cthoyt** [comment]
I added some additional logging and error handling to the tests, now it's able to pick up when configuration isn't correctly configured (which seems to be the case at the moment)

**Comment by cthoyt** [comment]
@nuest can you please break this into two separate PRs:

1. That changes functionality (i.e. for toggling `publish=False`)
2. That adds new functionality (i.e., `update_metadata()`)

The way I suggest doing this is just removing the `update_metadata()` functionality from this PR and making a separate one later.

The reason I'm asking this is there are too many changes that are hard to understand and read in the diff

**Comment by nuest** [comment]
I can do that. Just let me know when you are done with your edits - thanks!

**Comment by cthoyt** [comment]
@nuest all done, it's in your court :)

**Comment by nuest** [comment]
:tennis: Thanks! I'll get to it as early as possible next week.

My approach would be to make new branches from main and transferring the changes manually there - the git history of this PR is beyond my skillset to "pick" from it.

**Comment by cthoyt** [comment]
Easier - branch off of this branch, delete the stuff you don't need

**Comment by nuest** [comment]
Replaced by #28 and #29

**Pull request closed without merging**